src/backend/allocator/dmabuf.rs

```diff
@@ -251,6 +253,141 @@ impl Dmabuf {
         let blocker = DmabufBlocker(source.signal.clone());
         Ok((blocker, source))
     }
+
+    /// Map this dmabuf readable
+    pub fn map_readable(&self) -> Result<DmabufMapping, DmabufMappingFailed> {
+        self.map(rustix::mm::ProtFlags::READ).map(DmabufMapping)
+    }
+
+    /// Map this dmabuf read/writeable
+    pub fn map_writable(&self) -> Result<DmabufMappingMut, DmabufMappingFailed> {
```

Comment:
So... does this violate rust's safety constraints? Or would the mmap call error out mapping the same region twice?

Reply:
Creating `MAP_SHARED` mmaps of fds in Rust is only safe if either
* The fd is sealed against writes, or we trust any process with access to the fd not to write to it
* A slice is never created from the pointer

If the client modifies the contents of the dmabuf or shm fd while we have a slice into it, that is, at least technically, undefined behavior.

Reply:
> So... does this violate rust's safety constraints? Or would the mmap call error out mapping the same region twice?

Why would it error out when mapping the same fd twice, it creates a mmap for each call? Regarding the safety constraints, maybe. But if we want to map it mutable there is no way around that imo as we already have `Clone` on `Dmabuf`. So I did not feel the need to make the API more complicated.

> Creating MAP_SHARED mmaps of fds in Rust is only safe if either
>
>    The fd is sealed against writes, or we trust any process with access to the fd not to write to it
>    A slice is never created from the pointer
>
> If the client modifies the contents of the dmabuf or shm fd while we have a slice into it, that is, at least technically, undefined behavior.

In the pixman renderer it is actually easier to use the ptr to initialize a image, so I am fine with not providing a slice at all. But we need `MAP_SHARED` for writing to the dmabuf or otherwise kms won't see our changes. For the read-only case we could use `MAP_PRIVATE`.

Reply:
> But if we want to map it mutable there is no way around that imo as we already have Clone on Dmabuf.

Right, because we can already have multiple references to the same dmabuf, we can't enforce a single mutable reference here. I was just hoping we could catch that at runtime at least.

> In the pixman renderer it is actually easier to use the ptr to initialize a image, so I am fine with not providing a slice at all.

Getting a pointer seems to be a better solution.